larq_compute_engine/mlir/transforms/prepare_patterns_common.td

```diff
@@ -4,9 +4,56 @@ include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops.td"
 include "larq_compute_engine/mlir/ir/lce_ops.td"
 include "larq_compute_engine/mlir/transforms/op_removal_patterns.td"
 
+class ConstantValue<string val> : AttrConstraint<CPred<"IsConstantValue($_self, " # val # ")">>;
 
-// This relies on implementation details of larq.math.sign. We should make
-// this more general in the future
+def CreateTFBroadcastToOp : NativeCodeCall<
+    "$_builder.create<TF::BroadcastToOp>("
+        "$0.getLoc(),"
+        "RankedTensorType::get("
+            "$0.getType().cast<RankedTensorType>().getShape(),"
+            "getElementTypeOrSelf($1.getType())),"
+        "$1,"
+        "$2)">;
+
+def CreateTFShapeOp : NativeCodeCall<
+    "$_builder.create<TF::ShapeOp>($0.getLoc(), $1, $2)">;
+
+// Base quantiser patterns that match the `tf.where` implementation of `ste_sign`.
+multiclass QuantDequantPatterns<Op SelectOp> {
+  def : Pat<(SelectOp:$select_op
+                $cond,
+                (ConstantOp ConstantValue<"1.0f">),
+                (ConstantOp ConstantValue<"-1.0f">)),
+            (LQ_DequantizeOp
+                (LQ_QuantizeOp
+                    (CreateTFBroadcastToOp
+                        $select_op,
+                        $cond,
+                        (CreateTFShapeOp
+                            $select_op,
+                            $cond,
```

Comment:
Does this extract the shape from `$select_op` or from `$cond`? I think the right behaviour would be to use the shape from `$select_op`.

Reply:
Ah I think you're right there, that's a mistake. For some reason the explicit broadcasting tests still pass, so it must still be working, but yes, will fix.

Reply:
Not sure why it's working though, but I think it's good to fix. We can probably also reduce the number of arguments from `CreateTFShapeOp` then.

Reply:
I haven't reduced the number of args, cause I think it's slightly nicer to use the identical name/construction as in TF, but don't feel too strongly about that.

Reply:
OK, then let's keep it like this. I just thought if we would also fix the the use 32-bit argument inside the utility function we the signature would match the OP as defined in tablegen and wasn't aware that this utility is borrowed from TF.

Reply:
OK, then let's keep it like this. I just thought if we would also fix the the use 32-bit argument inside the utility function we the signature would match the OP as defined in tablegen and wasn't aware that this utility is borrowed from TF.